wtxmgr/db.go

```diff
@@ -0,0 +1,1311 @@
+//   }
+//   if it.err != nil {
+//           // Handle error
+//   }
+//
+// The spentness of the credit is not looked up for performance reasons (because
+// for unspent credits, it requires another lookup in another bucket).  If this
+// is needed, it may be checked like this:
+//
+//   spent := existsRawUnminedInput(ns, it.ck) != nil
+type unminedCreditIterator struct {
+	c      walletdb.Cursor
+	prefix []byte
+	ck     []byte
+	cv     []byte
+	elem   CreditRecord
+	err    error
+}
+
+func makeUnminedCreditIterator(ns walletdb.Bucket, txHash *wire.ShaHash) unminedCreditIterator {
+	c := ns.Bucket(bucketUnminedCredits).Cursor()
+	return unminedCreditIterator{c: c, prefix: txHash[:]}
+}
+
+func (it *unminedCreditIterator) readElem() error {
+	index, err := fetchRawUnminedCreditIndex(it.ck)
+	if err != nil {
+		return err
+	}
+	amount, change, err := fetchRawUnminedCreditAmountChange(it.cv)
+	if err != nil {
+		return err
+	}
+
+	it.elem.Index = index
+	it.elem.Amount = amount
+	it.elem.Change = change
+	// Spent intentionally not set
+
+	return nil
+}
+
+func (it *unminedCreditIterator) next() bool {
+	if it.c == nil {
+		return false
+	}
+
+	if it.ck == nil {
+		it.ck, it.cv = it.c.Seek(it.prefix)
+	} else {
+		it.ck, it.cv = it.c.Next()
+	}
+	if !bytes.HasPrefix(it.ck, it.prefix) {
+		it.c = nil
+		return false
+	}
+
+	err := it.readElem()
+	if err != nil {
+		it.err = err
+		return false
+	}
+	return true
+}
+
+func (it *unminedCreditIterator) delete() error {
+	err := it.c.Delete()
+	if err != nil {
+		str := "failed to delete unmined credit"
+		return storeError(ErrDatabase, str, err)
+	}
+	return nil
+}
+
+// Outpoints spent by unmined transactions are saved in the unmined inputs
+// bucket.  This bucket maps between each previous output spent, for both mined
+// and unmined transactions, to the hash of the unmined transaction.
+//
+// The key is serialized as such:
+//
+//   [0:32]   Transaction hash (32 bytes)
+//   [32:36]  Output index (4 bytes)
+//
+// The value is serialized as such:
+//
+//   [0:32]   Transaction hash (32 bytes)
+
+func putRawUnminedInput(ns walletdb.Bucket, k, v []byte) error {
+	err := ns.Bucket(bucketUnminedInputs).Put(k, v)
+	if err != nil {
+		str := "failed to put unmined input"
+		return storeError(ErrDatabase, str, err)
+	}
+	return nil
+}
+
+func existsRawUnminedInput(ns walletdb.Bucket, k []byte) (v []byte) {
+	return ns.Bucket(bucketUnminedInputs).Get(k)
+}
+
+func deleteRawUnminedInput(ns walletdb.Bucket, k []byte) error {
+	err := ns.Bucket(bucketUnminedInputs).Delete(k)
+	if err != nil {
+		str := "failed to delete unmined input"
+		return storeError(ErrDatabase, str, err)
+	}
+	return nil
+}
+
+// upgradeManager opens the tx store using the specified namespace or creates
+// and initializes it if it does not already exist.  It also provides
+// facilities to upgrade the data in the namespace to newer versions.
+func upgradeManager(namespace walletdb.Namespace) error {
```

Comment:
Also, it looks like this was based on the old waddrmgr upgrade code.  It was changed a bit to explicitly separate the open/create functionality.  It now creates the buckets at creation time instead of lazily checking/creating them on open each time.  That approach also makes the upgrade process easier to handle since new buckets are created specifically in a certain version when doing upgrades and well be created by default for all newly managers at the current version.